### DISCUSSION TITLE
API for obtaining spans and traces not working

### DISCUSSION THREAD

**Discussion by CarlosGuemS** [kiali/kiali#5543] [General]
Hello, I'm running the [Online Boutique demo](https://github.com/GoogleCloudPlatform/microservices-demo) by Google in a minikube service mesh using Istio 1.15 and Kiali. My issue is that I want to use the API to observe spans in the application, but after following the instructions in the [kiali_api.md](https://github.com/kiali/kiali/blob/master/kiali_api.md) the query returns the following error:

GET 0.0.0.0:20001/kiali/api/namespaces/default/apps/frontend/spans

Status: 503 Service Unavailable
Size: 144 Bytes
Time: 14 ms
{
  "error": "Jaeger GRPC client, stream error: rpc error: code = Internal desc = failed when searching for traces: start and end time must be set"
}

I get the same error when trying to get traces with /traces. The issue is that the document doesn't specify how to specificy this start and end times. I tried using resonable guesses like "start" or "startTime", or parameters for other queries like "sinceTime" without success.

Can somebody tell me what I'm missing?

As for my enviroment there is little to say:
- I'm running the cluster in a VM with Ubuntu 18 LTS.
- The minikube cluster is running with:
  - 4 cores
  - 16 GBs of RAM
  - Kubernetes version 1.23.8
- Istio 1.15.1, and whatever kiali version thier repo contained

**Comment by aljesusg** [comment]
Hi, I moved this to a discussion because is not a bug.

https://github.com/kiali/kiali/blob/master/kiali_api.md is a documentation for developers, 
The API is for internal use of the application, if you go to the service/traces you will see a correct display of traces, you need to set the startTime and the endTime for this, by default the UI set this to one hour ago.

The API is not ready to be used from outside the UI.

**Comment by jmazzitelli** [comment]
I wasn't even aware we still produced that API markdown doc.

As Alberto says, keep in mind you are on your own if you attempt to use the Kiali API - it is not meant for outside consumption and can/will change from version to version with no promises of any kind of backward compatibility.

**Reply by jmazzitelli** [reply]
Here's a discussion at the time we ripped out the docs from kiali.io precisely because the API is not for public use: https://github.com/kiali/kiali/issues/3290#issuecomment-704224481

**Comment by CarlosGuemS** [comment]
I see, thanks for the clarification! My idea was to use the API for a scrapper running in the same VM, to automatically obtain some metrics while testing different loads in the application (similar to Locust, but recollecting more information). Do you recommend an alternative on how to handle this? This was going to be run in toy applications in a testing enviroment, not to be used in deployed real-life applications, so I don't really care that much if the API changes across versions.

**Reply by jshaughn** [reply]
Although not supported the API does not usually have breaking changes.  If you are OK with it and it's not for production use, I think it's OK to use.  By the way, Kiali does collect some of its own metrics in Prometheus.  I'm not sure if we are reporting on the interactions you are interested in, but maybe.

**Reply by CarlosGuemS** [reply]
Yeah, I try to collect from metrics from Prometheus too, but the barrier to figure out how the specific metric you are looking is called and how to query it properly in Prometheus is much higher than Kiali. Hence I started to depend more on Kiali's API as, even if doesn't support what I'm trying to do, I managed to progress much quicker using it than Prometheus.

Ultimately, I'll try to use Kiali's API as long it allows me to, and when it doesn't I'll dig deeper and see if I can query the jaeger collector directly. Thanks everyone for your replies!
